compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala

```diff
@@ -620,27 +620,128 @@ private class ExtractAPICollector(using Context) extends ThunkHolder {
     import scala.util.hashing.MurmurHash3
+    import core.Constants.*
+
+    val seenInlines = mutable.HashSet.empty[Symbol]
+
+    if inlineSym ne NoSymbol then
+      seenInlines += inlineSym // do not hash twice a recursive def
+
+    def nameHash(n: Name, initHash: Int): Int =
+      val h =
+        if n.isTermName then
+          MurmurHash3.mix(initHash, TermNameHash)
+        else
+          MurmurHash3.mix(initHash, TypeNameHash)
+
+      // The hashCode of the name itself is not stable across compiler instances
+      MurmurHash3.mix(h, n.toString.hashCode)
+    end nameHash
+
+    def typeHash(tp: Type, initHash: Int): Int =
+      // Go through `apiType` to get a value with a stable hash, it'd
+      // be better to use Murmur here too instead of relying on
+      // `hashCode`, but that would essentially mean duplicating
+      // https://github.com/sbt/zinc/blob/develop/internal/zinc-apiinfo/src/main/scala/xsbt/api/HashAPI.scala
+      // and at that point we might as well do type hashing on our own
+      // representation.
+      var h = initHash
+      tp match
+        case ConstantType(c) =>
+          h = constantHash(c, h)
+        case TypeBounds(lo, hi) =>
+          h = MurmurHash3.mix(h, apiType(lo).hashCode)
+          h = MurmurHash3.mix(h, apiType(hi).hashCode)
+        case tp =>
+          h = MurmurHash3.mix(h, apiType(tp).hashCode)
+      h
+    end typeHash
+
+    def constantHash(c: Constant, initHash: Int): Int =
+      var h = MurmurHash3.mix(initHash, c.tag)
+      c.tag match
+        case NullTag =>
+          // No value to hash, the tag is enough.
+        case ClazzTag =>
+          h = typeHash(c.typeValue, h)
+        case _ =>
+          h = MurmurHash3.mix(h, c.value.hashCode)
+      h
+    end constantHash
+
+    /**An inline method that calls another inline method will eventually inline the call
+     * at a non-inline callsite, in this case if the implementation of the nested call
+     * changes, then the callsite will have a different API, we should hash the definition
+     */
+    def inlineReferenceHash(ref: Symbol, rhs: Tree, initHash: Int): Int =
+      var h = initHash
+
+      def paramssHash(paramss: List[List[Symbol]], initHash: Int): Int = paramss match
+        case Nil :: paramss1 =>
+          paramssHash(paramss1, MurmurHash3.mix(initHash, EmptyParamHash))
+        case params :: paramss1 =>
+          var h = initHash
+          val paramsIt = params.iterator
+          while paramsIt.hasNext do
+            val param = paramsIt.next
+            h = nameHash(param.name, h)
+            h = typeHash(param.info, h)
+            if param.is(Inline) then
+              h = MurmurHash3.mix(h, InlineParamHash) // inline would change the generated code
+          paramssHash(paramss1, h)
+        case Nil =>
+          initHash
+      end paramssHash
+
+      h = paramssHash(ref.paramSymss, h)
+      h = typeHash(ref.info.finalResultType, h)
```

Comment:
Could we use apiDef(ref) instead which is already supposed to have a stable hash representing the method signature? (though looking at it, it might need to be expanded slightly to take into account the Inline flag on params)

Reply:
which means actually using `apiType(foo).hashCode` is also unstable because applied types have an array internally (`xsbti.api.Parameterized`)

Reply:
yes `apiDef` could be augmented then to add extra info for an inline hashing mode

Reply:
updated